.github/workflows/ci.yml

```diff
@@ -1,65 +1,86 @@
 name: CI
 on: [push, pull_request]
-jobs:
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+env:
+  DEFAULT_NODE_VERSION: 22.x
+
+jobs:
   lint:
+    name: Lint
     runs-on: ubuntu-latest
     steps:
-      - uses: actions/checkout@v2
-      - uses: actions/cache@v2
+      - name: Use Node.js ${{ env.DEFAULT_NODE_VERSION }}
+        uses: actions/setup-node@v4
         with:
-          path: '**/node_modules'
-          key: ${{ runner.os }}-lint-modules-${{ hashFiles('**/yarn.lock') }}
-      - uses: actions/setup-node@v2
+          node-version: ${{ env.DEFAULT_NODE_VERSION }}
+      - name: Enable corepack
+        run: corepack enable
+      - name: Ensure line endings are consistent
+        run: git config --global core.autocrlf input
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Load dependency cache
+        uses: actions/cache@v4
         with:
-          node-version: 20.x
-      - run: corepack enable
-      - run: yarn install
-      - run: yarn run lint
-
+          path: .yarn
+          key: ${{ runner.os }}-modules-${{ hashFiles('yarn.lock') }}
+      - name: Install dependencies
+        run: yarn install --immutable
+      - name: Run ESLint
+        run: yarn run lint
   test:
+    name: Test
+    needs: lint
     runs-on: ${{ matrix.os }}
     strategy:
       matrix:
-        os: [ubuntu-latest, windows-latest, macos-latest]
+        os:
+          - ubuntu-latest
+          - windows-latest
+          - macos-latest
         node-version:
           - 18.x
           - 20.x
-          - 21.x
+          - 22.x
     steps:
       - name: Use Node.js ${{ matrix.node-version }}
-        uses: actions/setup-node@v2
+        uses: actions/setup-node@v4
         with:
           node-version: ${{ matrix.node-version }}
-      - name: Enable CorePack
+      - name: Enable corepack
         run: corepack enable
       - name: Ensure line endings are consistent
         run: git config --global core.autocrlf input
-      - name: Check out repository
-        uses: actions/checkout@v2
-      - uses: actions/cache@v2
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Load dependency cache
+        uses: actions/cache@v4
         with:
-          path: '**/node_modules'
-          key: ${{ runner.os }}-test-modules-${{ hashFiles('**/yarn.lock') }}
+          path: .yarn
```

Comment:
Why not cache `node_modules` anymore?

Reply:
Because the linking step (to unpack the downloaded packages into `node_modules`) is faster than the download step, so it is more efficient to cache the download cache of Yarn (located in `.yarn`). That way, Yarn can skip the download step, which is most expensive, and proceed to just comparing its local cache against the server versions, and then do unpacking.

Reply:
But if we just cache `node_modules`, we don't have to 1) download, and 2) link. So it's more efficient to cache that, right?

Reply:
I changed it back to caching `node_modules` instead.